core/log/profiler_hook_summary.cpp

```diff
@@ -0,0 +1,502 @@
+}
+
+
+std::string fmt_avg_duration(duration d, int64 count)
+{
+    return fmt_duration(
+        std::chrono::duration_cast<std::chrono::nanoseconds, int64>(d).count() /
+        count);
+}
+
+
+std::string fmt_fraction(duration part, duration whole)
+{
+    std::stringstream ss;
+    ss << std::setprecision(1) << std::fixed;
+    ss << (double(part.count()) / double(whole.count())) * 100.0 << " %";
+    return ss.str();
+}
+
+
+template <std::size_t size>
+void print_table(const std::array<std::string, size>& headers,
+                 const std::vector<std::array<std::string, size>>& table,
+                 std::ostream& stream)
+{
+    std::array<std::size_t, size> widths;
+    for (int i = 0; i < widths.size(); i++) {
+        widths[i] = headers[i].size();
+    }
+    for (const auto& row : table) {
+        for (int i = 0; i < widths.size(); i++) {
+            widths[i] = std::max(widths[i], row[i].size());
+        }
+    }
+    for (int i = 0; i < widths.size(); i++) {
+        stream << '|';
+        const auto align1 = (widths[i] - headers[i].size()) / 2;
+        const auto align2 = (widths[i] - headers[i].size()) - align1;
+        stream << std::string(align1, ' ') << headers[i]
+               << std::string(align2, ' ');
+    }
+    stream << "|\n";
+    for (int i = 0; i < widths.size(); i++) {
+        stream << '|';
+        // right-align for Markdown, assuming widths[i] > 0
+        stream << std::string(widths[i] - 1, '-') << (i == 0 ? '-' : ':');
+    }
+    stream << "|\n";
+    for (const auto& row : table) {
+        for (int i = 0; i < widths.size(); i++) {
+            stream << '|';
+            stream << std::setw(widths[i]) << (i > 0 ? std::right : std::left)
+                   << row[i];
+        }
+        stream << "|\n";
+    }
+}
+
+
+void print_summary(std::ostream& stream, const std::string& name, summary& s)
+{
+    stream << name << '\n'
+           << "Overhead estimate " << fmt_duration(s.overhead) << '\n';
+    s.pop_all(stream);
+    std::sort(s.entries.begin(), s.entries.end(),
+              [](const auto& lhs, const auto& rhs) {
+                  // reverse-sort by inclusive total time
+                  return lhs.inclusive > rhs.inclusive;
+              });
+    std::vector<std::array<std::string, 6>> table;
+    std::array<std::string, 6> headers({" name ", " total ", " total (self) ",
+                                        " count ", " avg ", " avg (self) "});
+    for (const auto& entry : s.entries) {
+        table.emplace_back(std::array<std::string, 6>{
+            " " + entry.name + " ", " " + fmt_duration(entry.inclusive) + " ",
+            " " + fmt_duration(entry.exclusive) + " ",
+            " " + std::to_string(entry.count) + " ",
+            " " + fmt_avg_duration(entry.inclusive, entry.count) + " ",
+            " " + fmt_avg_duration(entry.exclusive, entry.count) + " "});
+    }
+    print_table(headers, table, stream);
+}
+
+
+struct nested_summary_tree {
```

Comment:
You would like to sort each summary by their time, so you need build another tree for it, right?

Reply:
that's the one half of it, yes. The other half is that the tree structure is only available through the `parent_id` entries to walk bottom-up, but to print the results, I need to be able to walk top-down.